<h1 align="center">  Desafio Lopes LABS </h1>

<h2 align="right"> Bruno Ramalho Furlan </h2>

**Objetivo:** Criar um modelo para estimar a qualidade do vinho.

<h3 align="justified">Importando dados </h3>

<p align="justified"> Verificando tipo de dados para cada coluna e quantidade de linhas e colunas </p>

In [1]:
import csv
import pandas as pd


df = pd.read_csv("winequality.csv", sep=";")
df.head(1000)



type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0    White            7.0              0.27         0.36           20.70   
1    White            6.3              0.30         0.34            1.60   
2    White            8.1              0.28         0.40            6.90   
3    White            7.2              0.23         0.32            8.50   
4    White            7.2              0.23         0.32            8.50   
..     ...            ...               ...          ...             ...   
995  White            7.8              0.27         0.34            1.60   
996  White            6.0              0.26         0.34            1.30   
997  White            6.1              0.24         0.27            9.80   
998  White            8.0              0.24         0.30           17.45   
999  White            7.6              0.21         0.60            2.10   

     chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0        0.045                 45.0                 170.0   1.0010  3.00   
1        0.049                 14.0                 132.0   0.9940  3.30   
2        0.050                 30.0                  97.0   0.9951  3.26   
3        0.058                 47.0                 186.0   0.9956  3.19   
4        0.058                 47.0                 186.0   0.9956  3.19   
..         ...                  ...                   ...      ...   ...   
995      0.046                 27.0                 154.0   0.9927  3.05   
996      0.046                  6.0                  29.0   0.9924  3.29   
997      0.062                 33.0                 152.0   0.9966  3.31   
998      0.056                 43.0                 184.0   0.9997  3.05   
999      0.046                 47.0                 165.0   0.9936  3.05   

     sulphates alcohol  quality  
0         0.45     8.8        6  
1         0.49     9.5        6  
2         0.44    10.1        6  
3         0.40     9.9        6  
4         0.40     9.9        6  
..         ...     ...      ...  
995       0.45    10.5        6  
996       0.63    10.4        5  
997       0.47     9.5        6  
998       0.50     9.2        6  
999       0.54    10.1        7  

[1000 rows x 13 columns]

In [2]:
print(df.shape)

(6497, 13)


In [3]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
type                    6497 non-null object
fixed acidity           6497 non-null float64
volatile acidity        6497 non-null float64
citric acid             6497 non-null float64
residual sugar          6497 non-null float64
chlorides               6497 non-null float64
free sulfur dioxide     6497 non-null float64
total sulfur dioxide    6497 non-null float64
density                 6497 non-null float64
pH                      6497 non-null float64
sulphates               6497 non-null float64
alcohol                 6497 non-null object
quality                 6497 non-null int64
dtypes: float64(10), int64(1), object(2)
memory usage: 660.0+ KB
None


<p align="justified"> Foi verificado que o tipo da coluna com a variavel álcool ("alcohol") está como object e não como float. Para isso foi realizada a alteração do tipo de coluna para float.</p>

In [4]:
df["alcohol"] = pd.to_numeric(df["alcohol"], errors="coerce")

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
type                    6497 non-null object
fixed acidity           6497 non-null float64
volatile acidity        6497 non-null float64
citric acid             6497 non-null float64
residual sugar          6497 non-null float64
chlorides               6497 non-null float64
free sulfur dioxide     6497 non-null float64
total sulfur dioxide    6497 non-null float64
density                 6497 non-null float64
pH                      6497 non-null float64
sulphates               6497 non-null float64
alcohol                 6457 non-null float64
quality                 6497 non-null int64
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


<h3 align="justified">Eliminando duplicatas e linhas com valores nulos </h3>
<p align="justified"> 
Eliminando dados duplicados ou que não apresentam valores em todas as colunas para uma estimação do modelo a partir de todos os parâmetros.</p>

In [6]:
import numpy as np
df = df.drop_duplicates()
df = df.dropna(how='any',axis=0) 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5288 entries, 0 to 6496
Data columns (total 13 columns):
type                    5288 non-null object
fixed acidity           5288 non-null float64
volatile acidity        5288 non-null float64
citric acid             5288 non-null float64
residual sugar          5288 non-null float64
chlorides               5288 non-null float64
free sulfur dioxide     5288 non-null float64
total sulfur dioxide    5288 non-null float64
density                 5288 non-null float64
pH                      5288 non-null float64
sulphates               5288 non-null float64
alcohol                 5288 non-null float64
quality                 5288 non-null int64
dtypes: float64(11), int64(1), object(1)
memory usage: 578.4+ KB


<h3 align="justified">Verificando a correlação dos dados com a qualidade do vinho (método de Pearson)</h3>



In [7]:
df.corrwith(df["quality"], axis=0, drop=False, method='pearson')

fixed acidity          -0.080232
volatile acidity       -0.266608
citric acid             0.098393
residual sugar         -0.057475
chlorides              -0.201844
free sulfur dioxide     0.054241
total sulfur dioxide   -0.049555
density                -0.026537
pH                      0.039920
sulphates               0.040824
alcohol                 0.469674
quality                 1.000000
dtype: float64

<p align="justified">As 3 variáveis que apresentam maior correlação encontradas foram:</p>
    
1. Álcool ("alcohol"), correlação positiva (0.469674); 
2. Volatilidade da acidez ("volatile acidity"), correlação negativa (-0.266608);
3. Cloretos ("chlorides"), correlação negativa (-0.201844);

<p align="justified">Estas 3 variáveis apresentaram correlação fraca, as demais apresentaram correlação inferior, em módulo, a 0,1.</p> 
<p align="justified">Para fins de comparação, feitas as correlações utilizando outros 2 métodos de correlação (Spearman e Kendall), sendo selecionadas  as 3 maiores correlações em módulo.</p>
    
     

<h3 align="justified">Verificando a correlação dos dados com a qualidade do vinho (método de Spearman)</h3>


In [8]:
df.corrwith(df["quality"], axis=0, drop=False, method='spearman')

fixed acidity          -0.104139
volatile acidity       -0.252981
citric acid             0.116741
residual sugar         -0.029466
chlorides              -0.303533
free sulfur dioxide     0.090431
total sulfur dioxide   -0.057063
density                -0.349593
pH                      0.053313
sulphates               0.034335
alcohol                 0.479853
quality                 1.000000
dtype: float64

<p align="justified">As 3 variáveis que apresentam maior correlação encontradas foram:</p>
    
1. Álcool ("alcohol"), correlação positiva (0.479853); 
2. Densidade ("density"), correlação negativa (-0.349593);
3. Cloretos ("chlorides"), correlação negativa (-0.303533);


    

<h3 align="justified">Verificando a correlação dos dados com a qualidade do vinho (método de Kendall)</h3>


In [9]:
df.corrwith(df["quality"], axis=0, drop=False, method='kendall')

fixed acidity          -0.080642
volatile acidity       -0.194831
citric acid             0.090924
residual sugar         -0.022765
chlorides              -0.235499
free sulfur dioxide     0.069322
total sulfur dioxide   -0.044525
density                -0.268305
pH                      0.041284
sulphates               0.027102
alcohol                 0.377853
quality                 1.000000
dtype: float64

<p align="justified">As 3 variáveis que apresentam maior correlação encontradas foram:</p>
    
1. Álcool ("alcohol"), correlação positiva (0.377853); 
2. Densidade ("density"), correlação negativa (-0.268305);
3. Cloretos ("chlorides"), correlação negativa (-0.235499);

<p align="justified">Estas 3 variáveis apresentaram correlação fraca, as demais apresentaram correlação despresível (inferior, em módulo a 0,1).</p>
<p align="justified">Como o método de Kendall apresentou menor correlação das variáveis em comparação com os outros métodos, especialmente na variável álcool (que apresenta menor correlação em comparação com os demais métodos). Este não será utilizado na modelagem da qualidade. Assim ,serão comarados os métodos de Spearman e Pearson.</p>

<h3 align="justified">Verificando dados para a modelagem </h3>


In [10]:
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    5288.000000       5288.000000  5288.000000     5288.000000   
mean        7.219696          0.344149     0.318678        5.050662   
std         1.319125          0.168253     0.147339        4.506067   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.240000        1.800000   
50%         7.000000          0.300000     0.310000        2.700000   
75%         7.700000          0.410000     0.400000        7.500000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  5288.000000          5288.000000           5288.000000  5288.000000   
mean      0.056739            30.018154            114.042549     1.678458   
std       0.036944            17.825880             56.849174     7.517368   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            16.000000             74.000000     0.992200   
50%       0.047000            28.000000            116.000000     0.994695   
75%       0.066000            41.000000            153.000000     0.996782   
max       0.611000           289.000000            440.000000   103.898000   

                pH    sulphates      alcohol      quality  
count  5288.000000  5288.000000  5288.000000  5288.000000  
mean      3.224707     0.533264    10.546874     5.794629  
std       0.160434     0.149795     1.186073     0.880721  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.117500     0.430000     9.500000     5.000000  
50%       3.210000     0.510000    10.400000     6.000000  
75%       3.330000     0.600000    11.400000     6.000000  
max       4.010000     2.000000    14.900000     9.000000

<p align="justified"> Para para a modelagem nos dois métodos foram utilizados os valores máximo e mínimo das seguintes variáveis, além da variável de qualidade (utilizada para validar o modelo):</p>

1. Álcool;
2. Volatilidade da acidez;
3. Cloretos;
4. Densidade;
   
    
<p align="justified"> Para o método de Pearson foram utilizadas as variáveis álcool, volatilidade de acidez e cloretos enquanto que para o método de Spearman foram utilizadas as variáveis álcool, densidade e cloretos. para cada uma destas variáveis, foram dados conceitos de qualidade de 0 a 10 (para correlação positiva) e de 10 a 0 (para correlação negativa). Como o valor máximo de qualidade encontrado foi de 9 e o mínimo foi de 3, foram feitos os seguintes métodos para a estimação dos conceitos das variáveis:</p>


* Foi feita a subtração do menor valor para o maior valor de cada uma das váriáveis, pelo menor valor e este foi dividido por 6 (intervalo entre o maior e o menor valor da variável qualidade) para determinar o tamanho do intervalo de dados que receberia cada conceito;
* Com o tamanho do intervalo de dados para cada variável, foram feitas as divisões de intervalos para variável e atribuidos conceitos para cada variável:de 3 a 9 para correlações positivas e de 9 a 3 para correlações negativas;
* Com a utilização de cada intervalo de dados, foram estimados os possíveis intervalos para os conceitos de 0,1,2,3 e 10;


<p align="justified">Assim, voram estimados os seguintes intervalos e conceitos para cada variável</p>

1. Álcool;

| Intervalo |       | Conceito |
|-----------|-------|----------|
| Início    | Fim   |          |
| 0         | 5,7   | 0        |
| 5,7       | 6,85  | 1        |
| 6,85      | 8     | 2        |
| 8         | 9,15  | 3        |
| 9,15      | 10,3  | 4        |
| 10,3      | 11,45 | 5        |
| 11,45     | 12,6  | 6        |
| 12,6      | 13,75 | 7        |
| 13,75     | 14,9  | 8        |
| 14,9      | 16,05 | 9        |
| <= 16,05  |       | 10       |

2. Volatilidade da acidez;

| Intervalo |      | Conceito |
|-----------|------|----------|
| Início    | Fim  |          |
| 0         | 0,08 | 10       |
| 0,08      | 0,33 | 9        |
| 0,33      | 0,58 | 8        |
| 0,58      | 0,83 | 7        |
| 0,83      | 1,08 | 6        |
| 1,08      | 1,33 | 5        |
| 1,33      | 1,58 | 4        |
| 1,58      | 1,83 | 3        |
| 1,83      | 2,08 | 2        |
| 2,08      | 2,33 | 1        |
| <= 2,33   |      | 0        |

3. Cloretos;

| Intervalo |        | Conceito |
|-----------|--------|----------|
| Início    | Fim    |          |
| 0,0000    | 0,0090 | 10       |
| 0,0090    | 0,1093 | 9        |
| 0,1093    | 0,2097 | 8        |
| 0,2097    | 0,3100 | 7        |
| 0,3100    | 0,4103 | 6        |
| 0,4103    | 0,5107 | 5        |
| 0,5107    | 0,6110 | 4        |
| 0,6110    | 0,7113 | 3        |
| 0,7113    | 0,8117 | 2        |
| 0,8117    | 0,9120 | 1        |
| <= 0,9120 |        | 0        |

4. Densidade;

| Intervalo |          | Conceito |
|-----------|----------|----------|
| Início    | Fim      |          |
| 0,0000    | 0,9871   | 10       |
| 0,9871    | 18,1389  | 9        |
| 18,1389   | 35,2907  | 8        |
| 35,2907   | 52,4426  | 7        |
| 52,4426   | 69,5944  | 6        |
| 69,5944   | 86,7462  | 5        |
| 86,7462   | 103,8980 | 4        |
| 103,8980  | 121,0498 | 3        |
| 121,0498  | 138,2016 | 2        |
| 138,2016  | 155,3534 | 1        |
| <= 155,353|          | 0        |


<p align="justified">* O final de cada intervalo é aberto</p>
<p align="justified"> Em seguida, é feita uma cópia do banco de dados para cada método, sendo calculados os conceitos para cada variável em cada vinho, sendo também feita as sua qualidade estimada. O cálculo da qualidade estimada, é feito  pela média ponderada, multiplicando cada conceito de cada variável pelo módulo da correlação da mesma e dividindo pela soma do módulo das correlações. </p>

<p align="justified"> O erro de estimação também foi calculado para cada vinho, sendo este o módulo da diferença da qualidade no banco da dados e da qualidade estimada no modelo. Assim, para a determinação do modelo mais eficaz foi considerado o que apresentou menor média de erro de estimação dentre os dois métodos e por tipo de vinho.</p>

In [11]:
dfp=df.copy()
dfs=df.copy()

<h3 align="justified">Criando o modelo (método de Pearson)</h3>

In [12]:
dfp.insert(13, 'qualidade_alcool', 0)
dfp.insert(14, 'qualidade_volatilidade da acidez', 0)
dfp.insert(15, 'qualidade_cloretos', 0)
dfp.insert(16, 'qualidade_estimada', 0)
dfp.insert(17, 'erro', 0)
dfp.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    5288.000000       5288.000000  5288.000000     5288.000000   
mean        7.219696          0.344149     0.318678        5.050662   
std         1.319125          0.168253     0.147339        4.506067   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.240000        1.800000   
50%         7.000000          0.300000     0.310000        2.700000   
75%         7.700000          0.410000     0.400000        7.500000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  5288.000000          5288.000000           5288.000000  5288.000000   
mean      0.056739            30.018154            114.042549     1.678458   
std       0.036944            17.825880             56.849174     7.517368   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            16.000000             74.000000     0.992200   
50%       0.047000            28.000000            116.000000     0.994695   
75%       0.066000            41.000000            153.000000     0.996782   
max       0.611000           289.000000            440.000000   103.898000   

                pH    sulphates      alcohol      quality  qualidade_alcool  \
count  5288.000000  5288.000000  5288.000000  5288.000000            5288.0   
mean      3.224707     0.533264    10.546874     5.794629               0.0   
std       0.160434     0.149795     1.186073     0.880721               0.0   
min       2.720000     0.220000     8.000000     3.000000               0.0   
25%       3.117500     0.430000     9.500000     5.000000               0.0   
50%       3.210000     0.510000    10.400000     6.000000               0.0   
75%       3.330000     0.600000    11.400000     6.000000               0.0   
max       4.010000     2.000000    14.900000     9.000000               0.0   

       qualidade_volatilidade da acidez  qualidade_cloretos  \
count                            5288.0              5288.0   
mean                                0.0                 0.0   
std                                 0.0                 0.0   
min                                 0.0                 0.0   
25%                                 0.0                 0.0   
50%                                 0.0                 0.0   
75%                                 0.0                 0.0   
max                                 0.0                 0.0   

       qualidade_estimada    erro  
count              5288.0  5288.0  
mean                  0.0     0.0  
std                   0.0     0.0  
min                   0.0     0.0  
25%                   0.0     0.0  
50%                   0.0     0.0  
75%                   0.0     0.0  
max                   0.0     0.0

In [13]:
for index, row in dfp.iterrows():
    if row['alcohol']<5.7:
        dfp.at[index,'qualidade_alcool']=0
    elif row['alcohol']<6.85:
        dfp.at[index,'qualidade_alcool']=1
    elif row['alcohol']<8:
        dfp.at[index,'qualidade_alcool']=2
    elif row['alcohol']<9.15:
        dfp.at[index,'qualidade_alcool']=3
    elif row['alcohol']<10.30:
        dfp.at[index,'qualidade_alcool']=4
    elif row['alcohol']<11.45:
        dfp.at[index,'qualidade_alcool']=5
    elif row['alcohol']<12.60:
        dfp.at[index,'qualidade_alcool']=6
    elif row['alcohol']<13.75:
        dfp.at[index,'qualidade_alcool']=7
    elif row['alcohol']<14.90:
        dfp.at[index,'qualidade_alcool']=8
    elif row['alcohol']<16.05:
        dfp.at[index,'qualidade_alcool']=9
    elif row['alcohol']>16.05:
        dfp.at[index,'qualidade_alcool']=10

In [14]:
for index, row in dfp.iterrows():
    if row['volatile acidity']<0.08:
        dfp.at[index,'qualidade_volatilidade da acidez']=10
    elif row['volatile acidity']>=0.08:
        dfp.at[index,'qualidade_volatilidade da acidez']=9
    elif row['volatile acidity']>=0.33:
        dfp.at[index,'qualidade_volatilidade da acidez']=8
    elif row['volatile acidity']>=0.58:
        dfp.at[index,'qualidade_volatilidade da acidez']=7
    elif row['volatile acidity']>=0.83:
        dfp.at[index,'qualidade_volatilidade da acidez']=6
    elif row['volatile acidity']>=1.08:
        dfp.at[index,'qualidade_volatilidade da acidez']=5
    elif row['volatile acidity']>=1.33:
        dfp.at[index,'qualidade_volatilidade da acidez']=4
    elif row['volatile acidity']>=1.58:
        dfp.at[index,'qualidade_volatilidade da acidez']=3
    elif row['volatile acidity']>=1.83:
        dfp.at[index,'qualidade_volatilidade da acidez']=2
    elif row['volatile acidity']>=2.08:
        dfp.at[index,'qualidade_volatilidade da acidez']=1
    elif row['volatile acidity']>=2.33:
        dfp.at[index,'qualidade_volatilidade da acidez']=0

In [15]:
for index, row in dfp.iterrows():
    if row['chlorides']<0.009:
        dfp.at[index,'qualidade_cloretos']=10
    elif row['chlorides']>=0.009:
        dfp.at[index,'qualidade_cloretos']=9
    elif row['chlorides']>=0.1093:
        dfp.at[index,'qualidade_cloretos']=8
    elif row['chlorides']>=0.2097:
        dfp.at[index,'qualidade_cloretos']=7
    elif row['chlorides']>=0.3100:
        dfp.at[index,'qualidade_cloretos']=6
    elif row['chlorides']>=0.4103:
        dfp.at[index,'qualidade_cloretos']=5
    elif row['chlorides']>=0.5107:
        dfp.at[index,'qualidade_cloretos']=4
    elif row['chlorides']>=0.6110:
        dfp.at[index,'qualidade_cloretos']=3
    elif row['chlorides']>=0.7113:
        dfp.at[index,'qualidade_cloretos']=2
    elif row['chlorides']>=0.8117:
        dfp.at[index,'qualidade_cloretos']=1
    elif row['chlorides']>=0.9120:
        dfp.at[index,'qualidade_cloretos']=0

In [16]:
dfp['qualidade_estimada']=((dfp['qualidade_alcool']*0.47)+ (dfp['qualidade_volatilidade da acidez']*0.25)+(dfp['qualidade_cloretos']*0.2))/(0.47+0.25+0.2)
dfp['erro']=np.sqrt((dfp['qualidade_estimada']-dfp['quality'])**2)

In [17]:
dfp.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    5288.000000       5288.000000  5288.000000     5288.000000   
mean        7.219696          0.344149     0.318678        5.050662   
std         1.319125          0.168253     0.147339        4.506067   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.240000        1.800000   
50%         7.000000          0.300000     0.310000        2.700000   
75%         7.700000          0.410000     0.400000        7.500000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  5288.000000          5288.000000           5288.000000  5288.000000   
mean      0.056739            30.018154            114.042549     1.678458   
std       0.036944            17.825880             56.849174     7.517368   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            16.000000             74.000000     0.992200   
50%       0.047000            28.000000            116.000000     0.994695   
75%       0.066000            41.000000            153.000000     0.996782   
max       0.611000           289.000000            440.000000   103.898000   

                pH    sulphates      alcohol      quality  qualidade_alcool  \
count  5288.000000  5288.000000  5288.000000  5288.000000       5288.000000   
mean      3.224707     0.533264    10.546874     5.794629          4.740923   
std       0.160434     0.149795     1.186073     0.880721          1.060766   
min       2.720000     0.220000     8.000000     3.000000          3.000000   
25%       3.117500     0.430000     9.500000     5.000000          4.000000   
50%       3.210000     0.510000    10.400000     6.000000          5.000000   
75%       3.330000     0.600000    11.400000     6.000000          5.000000   
max       4.010000     2.000000    14.900000     9.000000          9.000000   

       qualidade_volatilidade da acidez  qualidade_cloretos  \
count                            5288.0              5288.0   
mean                                9.0                 9.0   
std                                 0.0                 0.0   
min                                 9.0                 9.0   
25%                                 9.0                 9.0   
50%                                 9.0                 9.0   
75%                                 9.0                 9.0   
max                                 9.0                 9.0   

       qualidade_estimada         erro  
count         5288.000000  5288.000000  
mean             6.824167     1.102147  
std              0.541913     0.697663  
min              5.934783     0.021739  
25%              6.445652     0.467391  
50%              6.956522     0.956522  
75%              6.956522     1.445652  
max              9.000000     4.978261

<h3 align="justified">Criando o modelo (método de Spearman)</h3>

In [18]:
dfs.insert(13, 'qualidade_alcool', 0)
dfs.insert(14, 'qualidade_densidade', 0)
dfs.insert(15, 'qualidade_cloretos', 0)
dfs.insert(16, 'qualidade_estimada', 0)
dfs.insert(17, 'erro', 0)
dfs.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    5288.000000       5288.000000  5288.000000     5288.000000   
mean        7.219696          0.344149     0.318678        5.050662   
std         1.319125          0.168253     0.147339        4.506067   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.240000        1.800000   
50%         7.000000          0.300000     0.310000        2.700000   
75%         7.700000          0.410000     0.400000        7.500000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  5288.000000          5288.000000           5288.000000  5288.000000   
mean      0.056739            30.018154            114.042549     1.678458   
std       0.036944            17.825880             56.849174     7.517368   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            16.000000             74.000000     0.992200   
50%       0.047000            28.000000            116.000000     0.994695   
75%       0.066000            41.000000            153.000000     0.996782   
max       0.611000           289.000000            440.000000   103.898000   

                pH    sulphates      alcohol      quality  qualidade_alcool  \
count  5288.000000  5288.000000  5288.000000  5288.000000            5288.0   
mean      3.224707     0.533264    10.546874     5.794629               0.0   
std       0.160434     0.149795     1.186073     0.880721               0.0   
min       2.720000     0.220000     8.000000     3.000000               0.0   
25%       3.117500     0.430000     9.500000     5.000000               0.0   
50%       3.210000     0.510000    10.400000     6.000000               0.0   
75%       3.330000     0.600000    11.400000     6.000000               0.0   
max       4.010000     2.000000    14.900000     9.000000               0.0   

       qualidade_densidade  qualidade_cloretos  qualidade_estimada    erro  
count               5288.0              5288.0              5288.0  5288.0  
mean                   0.0                 0.0                 0.0     0.0  
std                    0.0                 0.0                 0.0     0.0  
min                    0.0                 0.0                 0.0     0.0  
25%                    0.0                 0.0                 0.0     0.0  
50%                    0.0                 0.0                 0.0     0.0  
75%                    0.0                 0.0                 0.0     0.0  
max                    0.0                 0.0                 0.0     0.0

In [19]:
for index, row in dfs.iterrows():
    if row['alcohol']<5.7:
        dfs.at[index,'qualidade_alcool']=0
    elif row['alcohol']<6.85:
        dfs.at[index,'qualidade_alcool']=1
    elif row['alcohol']<8:
        df.at[index,'qualidade_alcool']=2
    elif row['alcohol']<9.15:
        dfs.at[index,'qualidade_alcool']=3
    elif row['alcohol']<10.30:
        dfs.at[index,'qualidade_alcool']=4
    elif row['alcohol']<11.45:
        dfs.at[index,'qualidade_alcool']=5
    elif row['alcohol']<12.60:
        dfs.at[index,'qualidade_alcool']=6
    elif row['alcohol']<13.75:
        dfs.at[index,'qualidade_alcool']=7
    elif row['alcohol']<14.90:
        dfs.at[index,'qualidade_alcool']=8
    elif row['alcohol']<16.05:
        dfs.at[index,'qualidade_alcool']=9
    elif row['alcohol']>16.05:
        dfs.at[index,'qualidade_alcool']=10

In [20]:
for index, row in dfs.iterrows():
    if row['density']<0.98711:
        dfs.at[index,'qualidade_densidade']=10
    elif row['density']>=0.98711:
        dfs.at[index,'qualidade_densidade']=9
    elif row['density']>=18.138925:
        dfs.at[index,'qualidade_densidade']=8
    elif row['density']>=35.29074:
        dfs.at[index,'qualidade_densidade']=7
    elif row['density']>=52.442555:
        dfs.at[index,'qualidade_densidade']=6
    elif row['density']>=69.59437:
        dfs.at[index,'qualidade_densidade']=5
    elif row['density']>=86.746185:
        dfs.at[index,'qualidade_densidade']=4
    elif row['density']>=103.898:
        dfs.at[index,'qualidade_densidade']=3
    elif row['density']>=121.049815:
        dfs.at[index,'qualidade_densidade']=2
    elif row['density']>=138.20163:
        dfs.at[index,'qualidade_densidade']=1
    elif row['density']>=155.353445:
        dfs.at[index,'qualidade_densidade']=0

In [21]:
for index, row in dfs.iterrows():
    if row['chlorides']<0.009:
        dfs.at[index,'qualidade_cloretos']=10
    elif row['chlorides']>=0.009:
        dfs.at[index,'qualidade_cloretos']=9
    elif row['chlorides']>=0.1093:
        dfs.at[index,'qualidade_cloretos']=8
    elif row['chlorides']>=0.2097:
        dfs.at[index,'qualidade_cloretos']=7
    elif row['chlorides']>=0.3100:
        dfs.at[index,'qualidade_cloretos']=6
    elif row['chlorides']>=0.4103:
        dfs.at[index,'qualidade_cloretos']=5
    elif row['chlorides']>=0.5107:
        dfs.at[index,'qualidade_cloretos']=4
    elif row['chlorides']>=0.6110:
        dfs.at[index,'qualidade_cloretos']=3
    elif row['chlorides']>=0.7113:
        dfs.at[index,'qualidade_cloretos']=2
    elif row['chlorides']>=0.8117:
        dfs.at[index,'qualidade_cloretos']=1
    elif row['chlorides']>=0.9120:
        dfs.at[index,'qualidade_cloretos']=0

In [22]:
dfs['qualidade_estimada']=((dfs['qualidade_alcool']*0.48)+(dfs['qualidade_densidade']*0.34)+(dfs['qualidade_cloretos']*0.30))/(0.48+0.34+0.30)
dfs['erro']=np.sqrt((dfs['qualidade_estimada']-dfs['quality'])**2)

In [23]:
dfs.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    5288.000000       5288.000000  5288.000000     5288.000000   
mean        7.219696          0.344149     0.318678        5.050662   
std         1.319125          0.168253     0.147339        4.506067   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.240000        1.800000   
50%         7.000000          0.300000     0.310000        2.700000   
75%         7.700000          0.410000     0.400000        7.500000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  5288.000000          5288.000000           5288.000000  5288.000000   
mean      0.056739            30.018154            114.042549     1.678458   
std       0.036944            17.825880             56.849174     7.517368   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            16.000000             74.000000     0.992200   
50%       0.047000            28.000000            116.000000     0.994695   
75%       0.066000            41.000000            153.000000     0.996782   
max       0.611000           289.000000            440.000000   103.898000   

                pH    sulphates      alcohol      quality  qualidade_alcool  \
count  5288.000000  5288.000000  5288.000000  5288.000000       5288.000000   
mean      3.224707     0.533264    10.546874     5.794629          4.740923   
std       0.160434     0.149795     1.186073     0.880721          1.060766   
min       2.720000     0.220000     8.000000     3.000000          3.000000   
25%       3.117500     0.430000     9.500000     5.000000          4.000000   
50%       3.210000     0.510000    10.400000     6.000000          5.000000   
75%       3.330000     0.600000    11.400000     6.000000          5.000000   
max       4.010000     2.000000    14.900000     9.000000          9.000000   

       qualidade_densidade  qualidade_cloretos  qualidade_estimada  \
count               5288.0              5288.0         5288.000000   
mean                   9.0                 9.0            7.174681   
std                    0.0                 0.0            0.454614   
min                    9.0                 9.0            6.428571   
25%                    9.0                 9.0            6.857143   
50%                    9.0                 9.0            7.285714   
75%                    9.0                 9.0            7.285714   
max                    9.0                 9.0            9.000000   

              erro  
count  5288.000000  
mean      1.413119  
std       0.728755  
min       0.142857  
25%       0.857143  
50%       1.285714  
75%       1.857143  
max       5.142857

<h3 align="justified">Comparando o erro de estimação nos diferentes tipos de vinho</h3>

In [24]:
dfp.groupby('type')['erro'].describe()

count      mean       std       min       25%       50%       75%  \
type                                                                        
Red    1354.0  1.201039  0.680783  0.021739  0.467391  1.445652  1.445652   
White  3934.0  1.068110  0.700240  0.021739  0.445652  0.956522  1.445652   

            max  
type             
Red    4.000000  
White  4.978261

In [25]:
dfs.groupby("type")['erro'].describe()

count      mean       std       min       25%       50%       75%  \
type                                                                        
Red    1354.0  1.535134  0.705848  0.142857  0.857143  1.714286  1.857143   
White  3934.0  1.371124  0.731870  0.142857  0.857143  1.285714  1.857143   

            max  
type             
Red    4.285714  
White  5.142857

<h3 align="justified">Conclusão</h3>

<p align="justified"> O erro de estimação médio  para o método de correlação de Pearson se mostrou menor que no método de Spearman, tanto para os valores totais quanto para o cada tipo de vinho , sendo este o mais adequado para  a modelagem.</p>